### PR TITLE
Add LatestEnvelopes and store & restore envelopes

### PR DESCRIPTION
**Pull request by TrustHenry**
It restores the node's latest nomination & ballot protocol messages
the next time the node boots up.
Store the latest envelopes in DB and restore to 'scp.setStateFromEnvelope'.

Related to #1170

### CODE REVIEW

**Comment by TrustHenry** [comment]
I have set sqlite as memory but I will replace it with a file.
I have an unknown error during the unit test.
Could you find a problem with my draftPR?

File: source/agora/consensus/LatestEnvelopes.d [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+import agora.common.ManagedDatabase;
+import agora.common.Serializer;
+import agora.common.Types;
+import agora.utils.Log;
+
+import d2sqlite3.library;
+import d2sqlite3.results;
+import d2sqlite3.sqlite3;
+
+import core.stdc.inttypes;
```
Comment (TrustHenry) [inline source/agora/consensus/LatestEnvelopes.d]: It can be removed now.

**Comment by AndrejMitrovic** [comment]
> I have an unknown error during the unit test.

Run your unittests with LLDB and when it segfaults type `bt` and hit enter.

File: source/agora/consensus/protocol/Nominator.d [hunk 2]
```diff
@@ -123,7 +128,8 @@ extern(D):
         this.scp = createSCP(this, node_id, IsValidator, no_quorum);
         this.taskman = taskman;
         this.ledger = ledger;
-        this.restoreSCPState(ledger);
+        this.restoreSCPState();
+        this.latest_envelopes = this.getLatestEnvelopes(":memory:");
```
Comment (AndrejMitrovic) [inline source/agora/consensus/protocol/Nominator.d]: I'll give you a hint: You are restoring SCP state, and then retrieving the latest envelopes after.

File: source/agora/consensus/protocol/Nominator.d [hunk 3]
```diff
@@ -451,8 +429,38 @@ extern(D):
             return;  // slot was already externalized, ignore outdated message
         }
 
-        if (this.scp.receiveEnvelope(envelope) != SCP.EnvelopeState.VALID)
+        if (this.scp.receiveEnvelope(envelope) == SCP.EnvelopeState.VALID)
+        {
+            vector!SCPEnvelope envelopes =
+                this.scp.getCurrentState(envelope.statement.slotIndex);
```
Comment (AndrejMitrovic) [inline source/agora/consensus/protocol/Nominator.d]: I thought about this more. I think it's inefficient if we do this every time we receive an envelope.

Instead we should do it:
- On node shutdown
- On node crash

For the first part I suggest adding an override of `shutdown` in the `Validator` class:

```
override void shutdown { }
{
    super.shutdown();
    this.nominator.storeLatestState();
}
```

And implementing `storeLatestState` which will store the envelopes to the database.

For node crashes we actually need to implement a crash / segfault event handler, however this is a separate problem and should not be solved as part of #1170.

Comment (TrustHenry) [reply source/agora/consensus/protocol/Nominator.d]: Yes, that's right.
I thought it was inefficient when I saw the frequency of calls.

File: source/agora/consensus/protocol/Nominator.d [hunk 4]
```diff
@@ -377,56 +383,28 @@ extern(D):
...
+            this.scp.setStateFromEnvelope(env.statement.slotIndex, env);
+            if (!this.scp.isSlotFullyValidated(env.statement.slotIndex))
+                assert(0);
+        }
 
         // there should at least be a genesis block
-        if (this.scp.empty())
-            assert(0);
+        // if (this.scp.empty())
+        //     assert(0);
```
Comment (TrustHenry) [inline source/agora/consensus/protocol/Nominator.d]: This verification is essential?

Comment (AndrejMitrovic) [reply source/agora/consensus/protocol/Nominator.d]: This can be removed. For the genesis block we won't have any state in SCP yet, and therefore this check is unnecessary.

File: source/agora/test/SCPState.d [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+    scope(failure) network.printLogs();
+    network.waitForDiscovery();
+
+    auto re_validator = network.clients[0];
+    auto txes = genesisSpendable().map!(txb => txb.sign()).array();
+    txes.each!(tx => re_validator.putTransaction(tx));
+    network.expectBlock(Height(1), 2.seconds);
+
+    // Now shut down & restart one node
+    network.restart(re_validator);
```
Comment (TrustHenry) [inline source/agora/test/SCPState.d]: I think I need to check more when node shutdown.

File: source/agora/consensus/LatestEnvelopes.d [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+    Copyright:
+        Copyright (c) 2020 BOS Platform Foundation Korea
+        All rights reserved.
+
+    License:
+        MIT License. See LICENSE for details.
+
+*******************************************************************************/
+
+module agora.consensus.LatestEnvelopes;
```
Comment (AndrejMitrovic) [inline source/agora/consensus/LatestEnvelopes.d]: `LatestEnvelope` is not a data type. So I think this is not a good name.

How about `agora.consensus.SCPEnvelopeDB` or `agora.consensus.SCPEnvelopeStore`?

File: source/agora/consensus/LatestEnvelopes.d [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+    public this (in string db_path)
+    {
+        const db_exists = db_path.exists;
+        if (db_exists)
+            log.info("Loading database from: {}", db_path);
+
+        this.db = new ManagedDatabase(db_path);
+
+        this.db.execute("CREATE TABLE IF NOT EXISTS store_envelopes " ~
+            "(seq INTEGER PRIMARY KEY AUTOINCREMENT, envelope BLOB NOT NULL)");
```
Comment (AndrejMitrovic) [inline source/agora/consensus/LatestEnvelopes.d]: This is wrong. This must be the slot index. It doesn't always start from zero. Don't use autoincrement.

File: source/agora/consensus/LatestEnvelopes.d [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+            SCPEnvelope envelope;
+            dg(envelope);
+        }
+        return 0;
+    }
+
+    /***************************************************************************
+
+        Returns:
+            the number of envelope in the latestEnvelopes
```
Comment (AndrejMitrovic) [inline source/agora/consensus/LatestEnvelopes.d]:
```suggestion
            the number of envelopes in the database
```

File: source/agora/consensus/protocol/Nominator.d [hunk 9]
```diff
@@ -455,6 +429,46 @@ extern(D):
...
+
+        Params:
+            data_dir = path to the data directory
+
+        Returns:
+            the latestEnvelopes instance
+
+    ***************************************************************************/
+
+    protected LatestEnvelopes getLatestEnvelopes (string data_dir)
```
Comment (AndrejMitrovic) [inline source/agora/consensus/protocol/Nominator.d]: This will also have to be renamed when you rename the module & class. Something like `getSCPEnvelopeDB`

File: source/agora/node/Validator.d [hunk 10]
```diff
@@ -360,4 +362,12 @@ public class Validator : FullNode, API
             this.enroll_man.updateRevealDistance(this.ledger.getBlockHeight());
         }
     }
+
+    /// Ditto
+    override void shutdown ()
+    {
+        super.shutdown();
+        const slot_idx = this.ledger.getBlockHeight() + 1;
```
Comment (AndrejMitrovic) [inline source/agora/node/Validator.d]: This is wrong. We should just get the latest state. Please make another commit where the slot index is stored in the Nominator as a field:

https://github.com/bpfkorea/agora/blob/e704747aec3c681976d50127ac507e15e524360c/source/agora/consensus/protocol/Nominator.d#L314

Then, you can just call `this.nominator.storeLatestState()` and all the logic will be contained within it.

File: source/agora/consensus/LatestEnvelopes.d [hunk 11]
```diff
@@ -0,0 +1,204 @@
+/*******************************************************************************
+
+    Contains supporting code for restoring the latest envelopes,
```
Comment (AndrejMitrovic) [inline source/agora/consensus/LatestEnvelopes.d]:
```suggestion
    Contains supporting code for restoring the latest SCP envelopes,
```

File: source/agora/consensus/LatestEnvelopes.d [hunk 12]
```diff
@@ -0,0 +1,204 @@
...
+
+        this.db = new ManagedDatabase(db_path);
+
+        this.db.execute("CREATE TABLE IF NOT EXISTS store_envelopes " ~
+            "(seq INTEGER PRIMARY KEY AUTOINCREMENT, envelope BLOB NOT NULL)");
+    }
+
+    /***************************************************************************
+
+        Add a envelope to the latestEnvelopes
```
Comment (AndrejMitrovic) [inline source/agora/consensus/LatestEnvelopes.d]:
```suggestion
        Store the envelope to the database.
````

File: source/agora/consensus/LatestEnvelopes.d [hunk 13]
```diff
@@ -0,0 +1,204 @@
...
+
+        First, clean with 'removeAll' before Adding it in new envelopes.
+
+        Params:
+            tx = the latestEnvelopes to add
```
Comment (AndrejMitrovic) [inline source/agora/consensus/LatestEnvelopes.d]: Wrong argument name `tx`

File: source/agora/consensus/protocol/Nominator.d [hunk 14]
```diff
@@ -455,6 +429,48 @@ extern(D):
...
+        this.latest_envelopes.removeAll();
+
+        // Store the lastest envelopes
+        foreach(const ref env; envelopes)
+            this.latest_envelopes.add(env);
+    }
+
+    /***************************************************************************
+
+        Returns an instance of a LatestEnvelopes
```
Comment (TrustHenry) [inline source/agora/consensus/protocol/Nominator.d]: I added the missing document.

File: source/agora/consensus/LatestEnvelopes.d [hunk 15]
```diff
@@ -0,0 +1,204 @@
...
+
+        Remove all envelopes from the latestEnvelopes
+
+    ***************************************************************************/
+
+    public void removeAll () @trusted nothrow
+    {
+        try
+        {
+            this.db.execute("DELETE FROM store_envelopes");
```
Comment (AndrejMitrovic) [inline source/agora/consensus/LatestEnvelopes.d]: Shouldn't this be:

```suggestion
            this.db.execute("DELETE * FROM store_envelopes");
```

I guess both work?

File: source/agora/consensus/LatestEnvelopes.d [hunk 16]
```diff
@@ -0,0 +1,204 @@
...
+
+    public this (in string db_path)
+    {
+        const db_exists = db_path.exists;
+        if (db_exists)
+            log.info("Loading database from: {}", db_path);
+
+        this.db = new ManagedDatabase(db_path);
+
+        this.db.execute("CREATE TABLE IF NOT EXISTS store_envelopes " ~
```
Comment (AndrejMitrovic) [inline source/agora/consensus/LatestEnvelopes.d]:
```suggestion
        this.db.execute("CREATE TABLE IF NOT EXISTS scp_envelopes " ~
```

File: source/agora/consensus/protocol/Nominator.d [hunk 17]
```diff
@@ -106,12 +111,13 @@ extern(D):
             key_pair = the key pair of this node
             ledger = needed for SCP state restoration & block validation
             taskman = used to run timers
+            data_dir = path to the data dirctory
```
Comment (AndrejMitrovic) [inline source/agora/consensus/protocol/Nominator.d]:
```suggestion
            data_dir = path to the data directory
```

File: source/agora/consensus/protocol/Nominator.d [hunk 18]
```diff
@@ -455,6 +429,48 @@ extern(D):
...
+            data_dir = path to the data directory
+
+        Returns:
+            the latestEnvelopes instance
+
+    ***************************************************************************/
+
+    protected LatestEnvelopes getLatestEnvelopes (string data_dir)
+    {
+        return new LatestEnvelopes(buildPath(data_dir, "store_envelopes.dat"));
```
Comment (AndrejMitrovic) [inline source/agora/consensus/protocol/Nominator.d]:
```suggestion
        return new LatestEnvelopes(buildPath(data_dir, "scp_envelopes.dat"));
```

File: source/agora/consensus/protocol/Nominator.d [hunk 19]
```diff
@@ -455,6 +429,48 @@ extern(D):
...
+
+    public void storeLatestState (ulong slot_index) @safe
+    {
+        vector!SCPEnvelope envelopes;
+        () @trusted
+        {
+            envelopes = this.scp.getCurrentState(slot_index);
+        }();
+
+        // Clean the lastest envelopes DB
```
Comment (AndrejMitrovic) [inline source/agora/consensus/protocol/Nominator.d]:
```suggestion
        // Clean the previous envelopes from the DB
```

File: source/agora/consensus/protocol/Nominator.d [hunk 20]
```diff
@@ -455,6 +429,48 @@ extern(D):
...
+        this.latest_envelopes.removeAll();
+
+        // Store the lastest envelopes
```
Comment (AndrejMitrovic) [inline source/agora/consensus/protocol/Nominator.d]:
```suggestion
        // Store the latest envelopes
```

File: source/agora/node/Validator.d [hunk 21]
```diff
@@ -280,9 +281,10 @@ public class Validator : FullNode, API
 
     protected Nominator getNominator (immutable(ConsensusParams) params,
         Clock clock, NetworkManager network, KeyPair key_pair, Ledger ledger,
-        TaskManager taskman)
+        TaskManager taskman, string data_dir)
```
Comment (AndrejMitrovic) [inline source/agora/node/Validator.d:287]: Needs parameter documentation.

File: source/agora/node/Validator.d [hunk 22]
```diff
@@ -360,4 +362,12 @@ public class Validator : FullNode, API
             this.enroll_man.updateRevealDistance(this.ledger.getBlockHeight());
         }
     }
+
+    /// Ditto
```
Comment (AndrejMitrovic) [inline source/agora/node/Validator.d]: What do you mean `/// Ditto`? That means `shutdown` becomes part of the `checkRevealPreimage` documentation because it's the previous symbol. Please document the function properly.

File: source/agora/test/Base.d [hunk 23]
```diff
@@ -440,6 +441,12 @@ extern(D):
 
         return true;
     }
+
+    /// Ditto
```
Comment (AndrejMitrovic) [inline source/agora/test/Base.d]: Wrong Ditto, please document this properly.

File: source/agora/test/SCPState.d [hunk 24]
```diff
@@ -0,0 +1,157 @@
+/*******************************************************************************
+
+    SCPState Related Test
```
Comment (AndrejMitrovic) [inline source/agora/test/SCPState.d]:
```suggestion
    Tests restoring SCP Envelope state on restart
```

File: source/agora/test/SCPState.d [hunk 25]
```diff
@@ -0,0 +1,157 @@
...
+    Copyright:
+        Copyright (c) 2020 BOS Platform Foundation Korea
+        All rights reserved.
+
+    License:
+        MIT License. See LICENSE for details.
+
+*******************************************************************************/
+
+module agora.test.SCPState;
```
Comment (AndrejMitrovic) [inline source/agora/test/SCPState.d]:
```suggestion
module agora.test.RestoreSCPState;
```

File: source/agora/test/SCPState.d [hunk 26]
```diff
@@ -0,0 +1,157 @@
...
+            TaskManager taskman, string data_dir, ulong txs_to_nominate)
+        {
+            super(params, clock, network, key_pair, ledger, taskman, data_dir,
+                txs_to_nominate);
+        }
+
+        ///
+        override void restoreSCPState ()
+        {
+            uint i;
```
Comment (AndrejMitrovic) [inline source/agora/test/SCPState.d]: You should add `assert(this.latest_envelopes.length > 0);` or `assert(i > 0)` after the foreach.

**Review by AndrejMitrovic — CHANGES_REQUESTED** [review]
First round of review done. I will review more after the fixes.

File: source/agora/consensus/LatestEnvelopes.d (hunk 6 above)
Comment (TrustHenry) [reply source/agora/consensus/LatestEnvelopes.d]: I've been thinking about a lot of names.
It's good name `agora.consensus.SCPEnvelopeStore`

File: source/agora/consensus/LatestEnvelopes.d (hunk 7 above)
Comment (TrustHenry) [reply source/agora/consensus/LatestEnvelopes.d]: `seq` is a meaningless unique key.
When checking the data in the envelope in the test code, the envelopes in the nodes in the same slot_index were identified.
If so, slot_idx cannot be used as a unique key.
slot_index is included in SCEnvelope.SCPStatement structure.
Is there a reason to use slot_index as a key?

Comment (AndrejMitrovic) [reply source/agora/consensus/LatestEnvelopes.d]: Ah it's in the statement. Yes you're right. Then it should be fine.

File: source/agora/consensus/LatestEnvelopes.d (hunk 15 above)
Comment (TrustHenry) [reply source/agora/consensus/LatestEnvelopes.d]: In standard SQL, most of the time,
Isn't it `DELETE FROM table name`?
Can't have * in a DELETE statement.

File: source/agora/node/Validator.d (hunk 10 above)
Comment (TrustHenry) [reply source/agora/node/Validator.d]: I was thinking about this.
Is there any way to get the latest slot_index in SCP.d?

Are these functions related?
https://github.com/bpfkorea/agora/blob/f7e514747f94a223b0d3ab99711220bd247e6657/source/scpd/scp/SCP.d#L117-L120

File: source/agora/consensus/protocol/Nominator.d (hunk 9 above)
Comment (TrustHenry) [reply source/agora/consensus/protocol/Nominator.d]: All names are being changed.

File: source/agora/node/Validator.d (hunk 10 above)
Comment (TrustHenry) [reply source/agora/node/Validator.d]: https://github.com/bpfkorea/agora/blob/e704747aec3c681976d50127ac507e15e524360c/source/agora/consensus/protocol/Nominator.d#L314
If I put this slot_idx in a variable, does it stay up to date?

Comment (AndrejMitrovic) [reply source/agora/node/Validator.d]: Yes you can use `getHighSlotIndex`.

**Comment by TrustHenry** [comment]
@AndrejMitrovic Thank you for your review.
Apply the suggestion now.
names and documents have been updated.
I have changed the reference place for slot_index.

File: source/agora/test/SCPState.d (hunk 26 above)
Comment (TrustHenry) [reply source/agora/test/SCPState.d]: Yes, there will be an error with the corresponding assert.
I'am looking for the cause.

File: source/agora/test/RestoreSCPState.d [hunk 27]
```diff
@@ -0,0 +1,159 @@
...
+            TaskManager taskman, string data_dir, ulong txs_to_nominate)
+        {
+            super(params, clock, network, key_pair, ledger, taskman, data_dir,
+                txs_to_nominate);
+        }
+
+        ///
+        override void restoreSCPState ()
+        {
+            //assert(this.scp_envelope_store.length > 0);
```
Comment (AndrejMitrovic) [inline source/agora/test/RestoreSCPState.d]: The reason it fails is because the node instantiates a new in-memory database when it restarts. It's empty.

Comment (TrustHenry) [reply source/agora/test/RestoreSCPState.d]: Yes, that's right.
I need to keep an instance.

Comment (TrustHenry) [reply source/agora/test/RestoreSCPState.d]: And `libc++abi.dylib: terminating with uncaught exception of type std::runtime_error: Cannot set state after starting ballot protocol`
Oh!...

Comment (TrustHenry) [reply source/agora/test/RestoreSCPState.d]: I don't know if the ballot protocol started.
I think need to make an exception.

Comment (AndrejMitrovic) [reply source/agora/test/RestoreSCPState.d]: I don't see this error anywhere in the CI.

Comment (AndrejMitrovic) [reply source/agora/test/RestoreSCPState.d]: And, you can only restore SCP state before nominations, before the event loop starts.

Comment (TrustHenry) [reply source/agora/test/RestoreSCPState.d]: I checked the message after testing it locally with FileDB.
I will keep the DB instance above and try to reproduce it.

**Comment by Geod24** [comment]
My understanding is that this PR still needs work and will carry over to the next sprint, correct ?

File: source/agora/consensus/protocol/Nominator.d [hunk 28]
```diff
@@ -466,6 +441,48 @@ extern(D):
...
+
+    public void storeLatestState () @safe nothrow
+    {
+        vector!SCPEnvelope envelopes;
+
+        () @trusted
+        {
+            if (this.scp.empty())
+                return;
+            envelopes = this.scp.getLatestMessagesSend(this.scp.getHighSlotIndex());
```
Comment (TrustHenry) [inline source/agora/consensus/protocol/Nominator.d:460]: Changed from getCurrentState() to getLatestMessagesSend()
getLatestMessagesSend is returns the latest messages sent for the given slot.

**Comment by TrustHenry** [comment]
> My understanding is that this PR still needs work and will carry over to the next sprint, correct ?

Yes, It is necessary to verify that SCP has implemented as intended.

File: source/agora/consensus/protocol/Nominator.d (hunk 28 above)
Comment (TrustHenry) [reply source/agora/consensus/protocol/Nominator.d:460]: The problem is not visible in this version.

File: source/agora/consensus/protocol/Nominator.d [hunk 29]
```diff
@@ -466,6 +441,48 @@ extern(D):
...
+        }();
+
+        // Clean the previous envelopes from the DB
+        this.scp_envelope_store.removeAll();
+
+        // Store the latest envelopes
+        foreach(const ref env; envelopes)
```
Comment (AndrejMitrovic) [inline source/agora/consensus/protocol/Nominator.d]:
```suggestion
        foreach (const ref env; envelopes)
```

File: source/agora/consensus/protocol/Nominator.d [hunk 30]
```diff
@@ -466,6 +441,48 @@ extern(D):
...
+
+        () @trusted
+        {
+            if (this.scp.empty())
+                return;
+            envelopes = this.scp.getLatestMessagesSend(this.scp.getHighSlotIndex());
+        }();
+
+        // Clean the previous envelopes from the DB
+        this.scp_envelope_store.removeAll();
```
Comment (AndrejMitrovic) [inline source/agora/consensus/protocol/Nominator.d:467]: I think this and the adding needs to be atomic with `db.begin` / `db.commit`. If removing succeeds but adding new envelopes fails, then we will have lost the old envelope messages.

File: source/agora/consensus/protocol/Nominator.d [hunk 31]
```diff
@@ -466,6 +441,48 @@ extern(D):
...
+
+        // Store the latest envelopes
+        foreach(const ref env; envelopes)
+            this.scp_envelope_store.add(env);
+    }
+
+    /***************************************************************************
+
+        Returns an instance of a SCPEnvelopeStore
```
Comment (AndrejMitrovic) [inline source/agora/consensus/protocol/Nominator.d]:
```suggestion
        Returns an instance of an SCPEnvelopeStore
```

(English is difficult, I know. :/
"A" and "An" is used based on pronunciation).

File: source/agora/consensus/protocol/Nominator.d [hunk 32]
```diff
@@ -466,6 +441,48 @@ extern(D):
...
+
+        Params:
+            data_dir = path to the data directory
+
+        Returns:
+            the latestEnvelopes instance
```
Comment (AndrejMitrovic) [inline source/agora/consensus/protocol/Nominator.d]:
```suggestion
            the SCPEnvelopeStore instance
```

File: source/agora/consensus/protocol/Nominator.d (hunk 28 above)
Comment (AndrejMitrovic) [reply source/agora/consensus/protocol/Nominator.d:460]: Just as a note to readers: We discussed this IRL. Stellar does the same thing here, using `getLatestMessagesSend()` instead of `getCurrentState`. It does mean that messages from other nodes are not recovered. However, the node will accept **new** messages from other nodes later when they are sent. It's only its own internal state that it needs to recover - IOW the node's own messages that it gossips.

This may not be very clear to people unless they understand SCP well.

**Comment by TrustHenry** [comment]
I did rebase done.

File: source/agora/consensus/SCPEnvelopeStore.d [hunk 33]
```diff
@@ -0,0 +1,199 @@
+/*******************************************************************************
+
+    Contains supporting code for restoring the latest SCP envelopes,
+    using SQLite as a backing store.
```
Comment (AndrejMitrovic) [inline source/agora/consensus/SCPEnvelopeStore.d]:
```suggestion
    Contains supporting code for storing the latest SCP envelopes,
    using SQLite as a backing store.
```

This code does not deal with restoring anything, just storing the data (being able to load it is implicit). The Nominator is in charge of doing the work of restoring.

**Review by AndrejMitrovic — CHANGES_REQUESTED** [review]

File: source/agora/consensus/protocol/Nominator.d (hunk 30 above)
Comment (TrustHenry) [reply source/agora/consensus/protocol/Nominator.d:467]: That's a good point.

File: source/agora/consensus/SCPEnvelopeStore.d (hunk 33 above)
Comment (TrustHenry) [reply source/agora/consensus/SCPEnvelopeStore.d]: Yes

File: source/agora/consensus/protocol/Nominator.d (hunk 29 above)
Comment (TrustHenry) [reply source/agora/consensus/protocol/Nominator.d]: I apply the suggestion.

File: source/agora/test/RestoreSCPState.d [hunk 34]
```diff
@@ -0,0 +1,171 @@
...
+    {
+        return store.length;
+    }
+}
+
+/// A test to store SCP state and recover SCP state
+unittest
+{
+    __gshared bool CheckEnvelopeStore = false;
+
```
Comment (AndrejMitrovic) [inline source/agora/test/RestoreSCPState.d:83]:
```suggestion
__gshared bool Checked = false;
```

File: source/agora/test/RestoreSCPState.d [hunk 35]
```diff
@@ -0,0 +1,171 @@
...
+        {
+            super(params, clock, network, key_pair, ledger, taskman, data_dir,
+                txs_to_nominate);
+        }
+
+        ///
+        override void restoreSCPState ()
+        {
+            if (CheckEnvelopeStore)
+                assert(TestSCPEnvelopeStore.store.length == 2);
```
Comment (AndrejMitrovic) [inline source/agora/test/RestoreSCPState.d]:
```suggestion
            // on first boot the envelope store will be empty,
            // but on restart it should contain two envelopes
            // (nominate & ballot protocol messages)
            if (CheckEnvelopeStore)
            {
                assert(TestSCPEnvelopeStore.store.length == 2);
                Checked = true;
            }
```

File: source/agora/test/RestoreSCPState.d [hunk 36]
```diff
@@ -0,0 +1,171 @@
...
+    auto re_validator = network.clients[0];
+    auto txes = genesisSpendable().map!(txb => txb.sign()).array();
+    txes.each!(tx => re_validator.putTransaction(tx));
+    network.expectBlock(Height(1), 2.seconds);
+
+    CheckEnvelopeStore = true;
+    // Now shut down & restart one node
+    network.restart(re_validator);
+    network.waitForDiscovery();
+    network.expectBlock(Height(1), 5.seconds);
```
Comment (AndrejMitrovic) [inline source/agora/test/RestoreSCPState.d:176]:
```suggestion
    network.expectBlock(Height(1), 5.seconds);
    assert(Checked);
```

**Review by AndrejMitrovic — CHANGES_REQUESTED** [review]
One final set of changes requested.

File: source/agora/test/RestoreSCPState.d [hunk 37]
```diff
@@ -0,0 +1,171 @@
...
+import geod24.Registry;
+
+import scpd.Cpp;
+import scpd.types.Stellar_SCP;
+
+import core.stdc.time;
+import core.thread;
+
+/// Class containing gshared store for SCPEnvelopeStoreTest
+public class TestSCPEnvelopeStore :  SCPEnvelopeStore
```
Comment (linked0) [inline source/agora/test/RestoreSCPState.d]:
```suggestion
public class TestSCPEnvelopeStore : SCPEnvelopeStore
```

File: source/agora/consensus/protocol/Nominator.d [hunk 38]
```diff
@@ -379,50 +387,18 @@ extern(D):
...
-                    commit: {
-                        counter: 0,
-                        value: value,
-                    },
-                    nH: 0,
-                },
-            },
-        };
-
-        SCPEnvelope env = SCPEnvelope(statement);
```
Comment (linked0) [inline source/agora/consensus/protocol/Nominator.d:418]: How about using the name of `envelope`. The `env` make me feel like 'environment'. But it depends on you.

**Review by linked0 — APPROVED** [review]
LGTM

File: source/agora/test/RestoreSCPState.d [hunk 39]
```diff
@@ -0,0 +1,178 @@
...
+        ///
+        override void restoreSCPState ()
+        {
+            // on first boot the envelope store will be empty,
+            // but on restart it should contain two envelopes
+            // (nominate & ballot protocol messages)
+            if (Checked)
+            {
+                assert(TestSCPEnvelopeStore.store.length == 2);
+                Checked = true;
```
Comment (AndrejMitrovic) [inline source/agora/test/RestoreSCPState.d]: There was some misunderstanding. I wanted you to use two booleans. However we can use a single boolean too. But, we need to change the code once more.

```suggestion
                Checked = false;
```

File: source/agora/test/RestoreSCPState.d [hunk 40]
```diff
@@ -0,0 +1,178 @@
...
+    auto txes = genesisSpendable().map!(txb => txb.sign()).array();
+    txes.each!(tx => re_validator.putTransaction(tx));
+    network.expectBlock(Height(1), 2.seconds);
+
+    Checked = true;
+    // Now shut down & restart one node
+    network.restart(re_validator);
+    network.waitForDiscovery();
+    network.expectBlock(Height(1), 5.seconds);
+    assert(Checked);
```
Comment (AndrejMitrovic) [inline source/agora/test/RestoreSCPState.d]:
```suggestion
    assert(!Checked);
```

**Review by AndrejMitrovic — CHANGES_REQUESTED** [review]
There was a misunderstanding.

File: source/agora/test/RestoreSCPState.d (hunk 39 above)
Comment (TrustHenry) [reply source/agora/test/RestoreSCPState.d]: Yes, that's right.

**Comment by TrustHenry** [comment]
@AndrejMitrovic @linked0  Thanks for the review.
Apply the suggestions.

**Review by AndrejMitrovic — APPROVED** [review]

**Comment by AndrejMitrovic** [comment]
> Related to #1170

@TrustHenry please use `Fixes #1170` in the PR description and the commit message so the associated issue is automatically closed in the future.
